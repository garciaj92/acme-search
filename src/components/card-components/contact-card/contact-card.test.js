import React from "react";
import ContactCard from "../contact-card";

import { mount, shallow } from "enzyme";
import toJson from 'enzyme-to-json';

// Mocks
import { consoleErrorFn } from "../../../__mocks__/console.mock";

const props = {
    name: "Jorge L. Garcia",
    company: "Neeva",
    emails: ["me@jorgegarcia.co"],
    phones: ["555-555-5555"],
    lastContact: "01/01/2020 - 4:00:00 PM"
};

describe("ContactCard: Test visual components", () => {
    let wrapper;
    beforeEach(() => {
        wrapper = mount(<ContactCard {...props} />);
    });

    it("renders correctly", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        // console.log(wrapper.debug());
    });

    it("displays the props correctly", () => {
        expect(wrapper.find("[data-testid='card-heading']").first().text()).toEqual(props.name);
        expect(wrapper.find("[data-testid='card-subheading']").first().text()).toEqual(props.company);
        expect(wrapper.find("[data-testid='email-info'] h5").first().text()).toEqual("Email");
        expect(wrapper.find("[data-testid='email-info'] a").first().text()).toEqual(props.emails[0]);
        expect(wrapper.find("[data-testid='phone-info'] h5").first().text()).toEqual("Phone");
        expect(wrapper.find("[data-testid='phone-info'] a").first().text()).toEqual(props.phones[0]);
        expect(wrapper.find("[data-testid='card-footer'] h5").first().text()).toEqual("Last Contacted");
        expect(wrapper.find("[data-testid='card-footer'] p").first().text()).toEqual(props.lastContact);
    });
});

describe("ContactCard: Test prop type validation", () => {
    let consoleErrorSpy;
    beforeEach(() => {
        console.error = consoleErrorFn
        consoleErrorSpy = jest.spyOn(console, "error");
    });

    afterEach(() => {
        consoleErrorSpy.mockClear();
    });

    it("does not throw an error when props are provided", () => {
        shallow(<ContactCard {...props} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    it("does not throw an error when only name and lastContact props are provided", () => {
        const { name, lastContact } = props;
        shallow(<ContactCard name={name} lastContact={lastContact} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    // What is stated below applies to the remaining test cases in this describe block:
    //      console.error is called twice for each prop validator that fails because
    //      there are two components that use each failing prop. This means that if
    //      two props fail we have four errors.

    it("throws an error when company prop is not of type `string`", () => {
        const { name, lastContact } = props;
        shallow(<ContactCard name={name} lastContact={lastContact} company={1} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when name prop is not of type `string`", () => {
        const { lastContact } = props;
        shallow(<ContactCard name={0} lastContact={lastContact}/>);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when either emails or phones props are not arrays of type `string`", () => {
        const { name, lastContact } = props;
        shallow(<ContactCard name={name} lastContact={lastContact} emails={[0,1]} phones={[0,1]} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(4);
    });
});