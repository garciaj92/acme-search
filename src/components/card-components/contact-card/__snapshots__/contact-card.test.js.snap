// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactCard: Test visual components renders correctly 1`] = `
<ContactCard
  company="Neeva"
  emails={
    Array [
      "me@jorgegarcia.co",
    ]
  }
  lastContact="01/01/2020 - 4:00:00 PM"
  name="Jorge L. Garcia"
  phones={
    Array [
      "555-555-5555",
    ]
  }
>
  <Card
    bg="white"
    body={false}
    className="shadow"
  >
    <div
      className="shadow card bg-white"
    >
      <CardBody>
        <div
          className="card-body"
        >
          <CardHeading
            data-testid="card-heading"
            heading="Jorge L. Garcia"
          >
            <h4
              className="font-weight-bold"
            >
              Jorge L. Garcia
            </h4>
          </CardHeading>
          <CardSubHeading
            data-testid="card-subheading"
            heading="Neeva"
          >
            <h5
              className="text-muted mb-0"
            >
              Neeva
            </h5>
          </CardSubHeading>
          <EmailInfo
            data-testid="email-info"
            emails={
              Array [
                "me@jorgegarcia.co",
              ]
            }
          >
            <div
              className="d-sm-inline-block align-baseline-top m-3"
              data-testid="email-info"
            >
              <CardIconSubHeading
                heading="Email"
                iconSrc="email.svg"
              >
                <div
                  className="d-flex justify-content-center align-items-center my-2"
                >
                  <Image
                    fluid={false}
                    rounded={false}
                    roundedCircle={false}
                    src="email.svg"
                    thumbnail={false}
                    width={30}
                  >
                    <img
                      className=""
                      src="email.svg"
                      width={30}
                    />
                  </Image>
                  <h5
                    className="mb-0 mx-1"
                  >
                    Email
                  </h5>
                </div>
              </CardIconSubHeading>
              <CardLink
                content="me@jorgegarcia.co"
                href="mailto:me@jorgegarcia.co"
                key="0"
              >
                <a
                  className="text-muted mb-0 d-block"
                  href="mailto:me@jorgegarcia.co"
                  onClick={[Function]}
                >
                  me@jorgegarcia.co
                </a>
              </CardLink>
            </div>
          </EmailInfo>
          <PhoneInfo
            data-testid="phone-info"
            phones={
              Array [
                "555-555-5555",
              ]
            }
          >
            <div
              className="d-sm-inline-block align-baseline-top mx-3"
              data-testid="phone-info"
            >
              <CardIconSubHeading
                heading="Phone"
                iconSrc="phone.svg"
              >
                <div
                  className="d-flex justify-content-center align-items-center my-2"
                >
                  <Image
                    fluid={false}
                    rounded={false}
                    roundedCircle={false}
                    src="phone.svg"
                    thumbnail={false}
                    width={30}
                  >
                    <img
                      className=""
                      src="phone.svg"
                      width={30}
                    />
                  </Image>
                  <h5
                    className="mb-0 mx-1"
                  >
                    Phone
                  </h5>
                </div>
              </CardIconSubHeading>
              <CardLink
                content="555-555-5555"
                href="tel:555-555-5555"
                key="0"
              >
                <a
                  className="text-muted mb-0 d-block"
                  href="tel:555-555-5555"
                  onClick={[Function]}
                >
                  555-555-5555
                </a>
              </CardLink>
            </div>
          </PhoneInfo>
        </div>
      </CardBody>
      <CardFooter
        className="bg-white"
        data-testid="card-footer"
      >
        <div
          className="bg-white card-footer"
          data-testid="card-footer"
        >
          <DateInfo
            date="01/01/2020 - 4:00:00 PM"
            title="Last Contacted"
          >
            <div
              className="text-muted"
            >
              <h5
                className="font-weight-bold mb-0"
              >
                Last Contacted
              </h5>
              <p
                className="font-weight-bold mb-0"
              >
                01/01/2020 - 4:00:00 PM
              </p>
            </div>
          </DateInfo>
        </div>
      </CardFooter>
    </div>
  </Card>
</ContactCard>
`;
